README.txt
Tortuga Branch
                          Memoranda version 1.0-RC3.1

================================================================================

Release notes
-------------

1.0-RC3.1 is a bugfix release for 1.0-RC3 (Release Candidate 3)  
  

Download
--------
 
Memoranda is distributed as a single zip-file containing the binaries, 
sourcecode and all dependency libraries.

The latest release can be downloaded from 
http://sourceforge.net/project/showfiles.php?group_id=90997

Contents
--------
Included in the release are the following:

    README.txt        this file 
    BUILD.txt      Building instructions
    FAQ.txt           Frequently Asked Questions 
    CHANGES.txt       Release changes log
    LICENSE           GPL software license
    memoranda.bat     Memoranda launch script for Win32 
    memoranda.sh      Memoranda launch script for Unix (bash)
    build.xml         Build configuration for Jakarta Ant
    build/            Contains the prebuilt 'memoranda.jar'
    lib/              Dependency libraries and other resources
    src/              Memoranda source code (not in the binary distribution)


System requirements
-------------------
For running Memoranda you need a Java 2 Runtime Environment (JRE) with a 
version number >= 1.4.0.  You can download the JRE for a number of platforms 
from http://java.sun.com. 

The "System tray icon" feature is available for 

    * MS Windows
    * K Desktop Environment (KDE) >3.0

Memoranda has been tested on Windows 2000, XP + j2re-1.4.0, 1.4.1, 1.4.2, 1.5
Linux + KDE + j2re-1.4.0, 1.4.1_01, 1.5.


Installation
------------
ALL PLATFORMS:
Unpack the downloaded zip file into a dir of your choice. It will 
create there the "memoranda**-**" subdirectory that will be known 
as the "Memoranda home dir".

IMPORTANT: JRE executables directory (jre/bin) must be on your system PATH.

*NIX:
Cd to Memoranda directory and execute:

chmod 755 ./memoranda.sh
chmod 755 ./lib/kde/systray4jd

KDE USERS: 
You optionally can put the systray4jd daemon in your system autostart and remove 
its startup line from memoranda.sh. 

NON-KDE USERS: 
You can safely remove the systray4j startup line from memoranda.sh.

Running
-------

    1) Go to the Memoranda home dir. 
    2) Run the 'memoranda.bat' (on Windows) or './memoranda.sh' (on Unix)
Enjoy!

NOTE: To launch Memoranda in minimized mode (system tray only), use the "-m" 
option. You can also set this behavior in File->Preferences menu.


Memoranda on the Net
--------------------

    * Memoranda homepage: http://memoranda.sf.net

    * Memoranda project on SourceForge: http://sf.net/projects/memoranda

Mailing lists
-------------

    * Main discussion list: memoranda-list@lists.sourceforge.net
      Subscribe: 
          https://lists.sourceforge.net/lists/listinfo/memoranda-list

        * Announce list: memoranda-announce@lists.sourceforge.net
      Subscribe: 
          https://lists.sourceforge.net/lists/listinfo/memoranda-announce

      
Contributors
------------

    * Alex Alishevskikh <alexeya@users.sourceforge.net>
    * Patrick Bielen    <pbielen(at)users.sourceforge.net>  
    * Ryan Ho <rawsushi(at)users.sourceforge.net>
    * Ivan Ribas   <ivanrise(at)msn.com>  
    * Jyrki Velhonoja <velhonoja (at) kapsi.fi>
    * Jeremy Whitlock <whitlock(at)starprecision.com>

If you want to get involved please email us.


Legal notices
-------------
Copyright (c) 2002-2007 Memoranda project team.
All rights reserved.

This product is licensed under the terms and conditions of the 
GNU General Public License (see LICENSE). NO WARRANTY.


Dependencies (./lib directory)
------------------------------

[lib/xom-1.0.jar]
   XOM (XML Object Model) API
    Copyright (c) 2003 Elliotte Rusty Harold. All Rights Reserved. 
    License: LGPL (see lib/xom.LICENSE)    
    Version: 1.0d21 
    Required

[lib/xmlParserAPIs.jar]
[lib/xercesImpl.jar]
   Apache Xerces XML parser
    Copyright (c) 1999-2002 The Apache Software Foundation. 
    All Rights Reserved. This product includes software developed 
    by the Apache Software Foundation (http://www.apache.org/).
    License: Apache Software License (see lib/xerces.LICENSE)
    Version: 2.1.0
    Required

[lib/systray4j.jar]
[lib/win32/systray4j.dll]
[lib/kde/systray4jd]
   SysTray for Java
    (c) 2003 SnoozeSoft (http://www.eikon.tum.de/~tamas/)
    License: LGPL (see lib/systray4j.LICENSE)
    Version: 2.4.1
    Required

[nekohtml.jar]
[nekohtmlXni.jar]
   CyberNeko HTML Parser
    (C) Copyright 2002-2003, Andy Clark. All rights reserved. 
    This product includes software developed by Andy Clark.
    License: The CyberNeko Software License, Version 1.0 
    (see lib/nekohtml.LICENSE)
    Version: 0.7.6
    Optional - required for XHTML export only.

[lib/kunststoff.jar]
   Kunststoff look-and-feel library
    Copyright (c) INCORS GmbH (www.incors.com)
    License: LGPL (see lib/kunststoff.LICENSE)
    Version: 2.0.1
    Optional

Some icons are taken from the Connectiva Crystal icon theme for KDE, 
(c) Everaldo (www.everaldo.com)
     
Thanks to everybody working on third-party products.

The Memoranda team, 
2007

================================================================================
$Id: README.txt,v 1.11 2007/04/06 21:21:21 alexeya Exp $

CHANGES.txt
Done since Memoranda 1.0-RC3 (20050517)

- Fixed: Dates in TaskTable were not localized
- Patch '[ 1681680 ] File upload as project ressource' applied
- Fixed: memoranda.bat
- Bug [1590113] is fixed partially (TaskTable tooltips are disabled over to-do labels)
- Fixed: [1571820] Spaces in resource names
- Fixed: [1531944] Checking tasklist DOCTYPE is removed

Done since Memoranda 1.0-RC2 (20050517)

- New task searching algorythm implemented (no XPath queries). The following
libraries are excluded: saxon8.jar, saxon8-xom.jar, nux.jar [alexeya] 
- Antialiasing of text in notes [alexeya]
- Editor fonts customization implemented (Preferences dialog) [alexeya]
- "Endless tasks" logic improved [alexeya]
- Fixed:NullPointer when LAST_OPENED_PROJECT_ID references to missing project [alexeya]
- Fixed:I18N: bug with unicode characters input (htmleditor) [alexeya]
- HTMLEditor code is moved into main codebase (net.sf.memoranda.ui.htmleditor
  package) [alexeya]
- ID generation algorythm implemented instead of uuidgen.jar library using [alexeya]
- TreeTable view for tasks hierarchy [alexeya, velhonoja]

Done since Memoranda 1.0-RC1 (20041109)

- Memoranda now remembers notes sort order
- Fixed: [ 1095685 ] Add Planned Task [ivanrise]
- Serbian messages added (submitted by Sava Rados <sawa@softhard.org>) [alexeya]
- Feature: Templates for HTML export [alexeya]
- Fixed: Overwriting output files with export of multiple notes per day [alexeya]
- Italian messages added (submitted by Milena Vitali-Charewicz <milo22370@yahoo.com>) [alexeya]
- German messages added (submitted by Michael Radtke <mradtke@abigale.de>)[alexeya]

Done since Memoranda 1.0b3-20040322
// On the homestretch before the Final Release:

- CHECK_IF_ALREADY_STARTED configuration option added, FAQ updated 
  (submitted by Michael Radtke <mradtke@abigale.de>)[alexeya]
- New feature: Preview a note with web-browser [alexeya]
- Icons updated [alexeya]
- Configurable port number (PORT_NUMBER property added)
  (see also FAQ.txt) [alexeya]
- Fixed: Incompatibility with old notes format. [alexeya]
- "File->New note" menu item added [alexeya]
- Fixed: fixed compile warnings on java 1.5
- Feature: delete notes using the DELETE key [ivanrise]
- Fixed: [ 1046374 ] not active selected notes [ivanrise]
- Fixed: Current note is not changed when another project is selected [alexeya]
- TaskPanel: "Complete task" icon added, "Show active only" implemented
  as checkbox in context menu (for consistency with projects menu) [alexeya] 
- Nicer color for day with notes in calendar [alexeya]
- Some configuration hack options added (memoranda.properties): 
    MEMORANDA_HOME - overrides default dir for projects data 
    ("${user.home}/.memoranda") 
    USER_CSS - overrides default CSS file for notes editor
    LOCALES_DIR - overrides default dir to search local messages      [alexeya] 
- New feature:  848119    Daily Events : Only Working days [ivanrise]
- New feature: [ 1042012 ] Change task sort order with "status" [ivanrise]
- We do not need the hoar's quicksort now. Tasks, Events and notes are sorted 
    using the Comparable interface.[ivanrise]
- Feature: [ 1039131 ] DEL deletes Project (projects/tasks/events/resources) [ivanrise]
- Fixed: [ 1039507 ] endless tasks displayed with wrong colors [ivanrise]
- Added new icons to the task panel
- Fixed: [ 1039443 ] calendar panel not updated [ivanrise]
- Fixed: [ 1039112 ] Translation Hints [ivanrise]
- Fixed: [ 899395 ] Memoranda crashes on JVM 1.5.0 [ivanrise]
- Fixed: [ 1038393 ] Wrong window dimensions [ivanrise]
- Fixed: [ 1038044 ] Problems with date selection [ivanrise]
- New feature: [ 934556 ] Multiple notes per project per day [ivanrise]
  At the moment there are no changes at the user interface. All notes ares
  displayed in the notelist. I just added a new button on the editor panel
  so we can create multiple notes on one day.
  -->> OLD NOTES ARE NOT COMPATIBLE!!! i changed the xml format as it was told in 
  the memoranda's developers forum.
- New feature:  848119    Daily Events : Only Working days [ivanrise]

- Fixed: Bug 988998: End date is not working on new projects dialog [ivanrise]
- End date on tasks default state is now not enabled. [ivanrise] 
- Request:843536 Event PopUp-Dialog(Reconstruct).It now looks better [ivanrise]
- New feature: added task descriptions [rawsushi]
- Show active tasks only will now show ACTIVE, DEADLINE and FAILED tasks instead of filtering by date [rawsushi]
- New feature: Complete task will automatically set the
  %done to 100%. (we need a new icon) [rawsushi / commit ivanrise]
- New feature: Show active tasks only (we need a new icon) [rawsushi / commit ivanrise]
- Fixed: Bug 965280: tasks with no end date are not displayed in agenda [ivanrise]
- default.css is fixed by I18N reasons (default serif and sans-serif fonts for text) [alexeya]
- Usability of dialog boxes: default buttons and style consistency [alexeya]
- Splash screen updated [alexeya]
- "About" dialog updated [alexeya]
- New Help menu items added ("Online user's guide", "Web site" and 
  "Report a bug") [alexeya] 
- All codebase : All catched exceptions are tracked by ExceptionDialog 
  (net.sf.memoranda.ui.ExceptionDialog.java) [alexeya]
- New feature: Request 934540:    Open-ended tasks [ivanrise]

Done since Memoranda 1.0b-20031127

- Chinese messages file added (messages_zh.properties) (submitted by <liu-610@163.com>) [alexeya]
- Current project title in the window titlebar [alexeyA]
- Fixed : Bug 892368 : Bug on moving TaskTable columns. [ivanrise]
- All codebase: Comment with CVS Id keyword added [alexeyA]
- build.xml: App.VERSION_INFO and App.BUILD_INFO values are set at buildtime [alexeyA]
- build.xml: Ant tasks for making the distributions changed [alexeyA]
- New feature: Stickers on the Agenda panel [alexeyA]
- New feature: Request 853579: Compile all notes in a single document [alexeyA]
- New feature: Sound event notifications [alexeyA]

src/net/sf/memoranda/ui/App.java
package net.sf.memoranda.ui;

import java.awt.Dimension;
import java.awt.Frame;
import java.awt.Toolkit;
import java.util.Calendar;

import javax.swing.ImageIcon;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.UIManager;

import net.sf.memoranda.EventNotificationListener;
import net.sf.memoranda.EventsScheduler;
import net.sf.memoranda.util.Configuration;
import net.sf.memoranda.util.Local;
import snoozesoft.systray4j.SysTrayMenu;
import snoozesoft.systray4j.SysTrayMenuEvent;
import snoozesoft.systray4j.SysTrayMenuIcon;
import snoozesoft.systray4j.SysTrayMenuItem;
import snoozesoft.systray4j.SysTrayMenuListener;
/**
 * 
 * Copyright (c) 2003 Memoranda Team. http://memoranda.sf.net
 */

/*$Id: App.java,v 1.28 2007/03/20 06:21:46 alexeya Exp $*/
public class App implements SysTrayMenuListener {
	// boolean packFrame = false;

	static AppFrame frame = null;
	static SysTrayMenu menu = null;

	static SysTrayMenuIcon icon = null;
	static SysTrayMenuIcon iconEv = null;
	static SysTrayMenuItem itemExit = null;
	static SysTrayMenuItem itemMinimize = null;
	static SysTrayMenuItem itemRestore = null;

	public static boolean sysTrayDisabled = true;
	
	public static final String GUIDE_URL = "http://memoranda.sourceforge.net/guide.html";
	public static final String BUGS_TRACKER_URL = "http://sourceforge.net/tracker/?group_id=90997&atid=595566";
	public static final String WEBSITE_URL = "http://memoranda.sourceforge.net";

	private JFrame splash = null;

	static {
		if (!Configuration.get("DISABLE_SYSTRAY").equals("yes")) {
			icon =
				new SysTrayMenuIcon(
					AppFrame.class.getResource(
						"resources/icons/jnotes"
							+ SysTrayMenuIcon.getExtension()));
			iconEv =
				new SysTrayMenuIcon(
					AppFrame.class.getResource(
						"resources/icons/jnotes_ev"
							+ SysTrayMenuIcon.getExtension()));
			itemExit = new SysTrayMenuItem(Local.getString("Exit"), "exit");
			itemMinimize =
				new SysTrayMenuItem(Local.getString("Minimize"), "minimize");
			itemRestore =
				new SysTrayMenuItem(Local.getString("Restore"), "restore");
			menu = new SysTrayMenu(icon, "Memoranda");
			sysTrayDisabled = false;
			if (!SysTrayMenu.isAvailable()) {
				System.out.println("[DEBUG] SysTray is now disabled.");
				Configuration.put("DISABLE_SYSTRAY", "yes");
				Configuration.put("ASK_ON_EXIT", "yes");
				Configuration.put("ON_CLOSE", "exit");
				Configuration.saveConfig();
				sysTrayDisabled = true;
			}
		} else
			/* DEBUG */
			System.out.println("System tray icon is disabled");
	}

	/*========================================================================*/ 
	/* Note: Please DO NOT edit the version/build info manually!
       The actual values are substituted by the Ant build script using 
       'version' property and datestamp.*/

	public static final String VERSION_INFO = "@VERSION@";
	public static final String BUILD_INFO = "@BUILD@";
	
	/*========================================================================*/

	public static AppFrame getFrame() {
		return frame;
	}

	public void show() {
		if (frame.isVisible()) {
			frame.toFront();
			frame.requestFocus();
		} else
			init();
	}

	public App(boolean fullmode) {
		super();
		if (fullmode)
			fullmode = !Configuration.get("START_MINIMIZED").equals("yes");
		/* DEBUG */
		if (!fullmode)
			System.out.println("Minimized mode");
		if (!Configuration.get("SHOW_SPLASH").equals("no"))
			showSplash();
		System.out.println(VERSION_INFO);
		try {
			if (Configuration.get("LOOK_AND_FEEL").equals("system"))
				UIManager.setLookAndFeel(
					UIManager.getSystemLookAndFeelClassName());
			else if (Configuration.get("LOOK_AND_FEEL").equals("kunststoff"))
				UIManager.setLookAndFeel(
					"com.incors.plaf.kunststoff.KunststoffLookAndFeel");
			else if (Configuration.get("LOOK_AND_FEEL").equals("default"))
				UIManager.setLookAndFeel(
					UIManager.getCrossPlatformLookAndFeelClassName());					
			else if (
				Configuration.get("LOOK_AND_FEEL").toString().length() > 0)
				UIManager.setLookAndFeel(
					Configuration.get("LOOK_AND_FEEL").toString());

		} catch (Exception e) {		    
			new ExceptionDialog(e, "Error when initializing a pluggable look-and-feel. Default LF will be used.", "Make sure that specified look-and-feel library classes are on the CLASSPATH.");
		}
		if (Configuration.get("FIRST_DAY_OF_WEEK").equals("")) {
			String fdow;
			if (Calendar.getInstance().getFirstDayOfWeek() == 2)
				fdow = "mon";
			else
				fdow = "sun";
			Configuration.put("FIRST_DAY_OF_WEEK", fdow);
			Configuration.saveConfig();
			/* DEBUG */
			System.out.println("[DEBUG] first day of week is set to " + fdow);
		}
		if (!sysTrayDisabled) {
			icon.addSysTrayMenuListener(this);
			iconEv.addSysTrayMenuListener(this);
			itemExit.addSysTrayMenuListener(this);
			itemMinimize.addSysTrayMenuListener(this);
			itemRestore.addSysTrayMenuListener(this);
			menu.addItem(itemExit, 0);
			menu.addSeparator(1);
			menu.addItem(itemRestore, 2);
			EventsScheduler.addListener(new EventNotificationListener() {
				public void eventIsOccured(net.sf.memoranda.Event ev) {
				}
				public void eventsChanged() {
					if (EventsScheduler.isEventScheduled()) {
						net.sf.memoranda.Event ev =
							EventsScheduler.getFirstScheduledEvent();
						menu.setToolTip(
							ev.getTimeString() + " - " + ev.getText());
						menu.setIcon(iconEv);
					} else {
						menu.setToolTip("Memoranda");
						menu.setIcon(icon);
					}
				}
			});
		}
		EventsScheduler.init();
		frame = new AppFrame();
		if (fullmode) {
			init();
		}
		if (!Configuration.get("SHOW_SPLASH").equals("no"))
			splash.dispose();
	}

	void init() {
		if (!sysTrayDisabled) {
			menu.removeAll();
			menu.addItem(itemExit, 0);
			menu.addSeparator(1);
			menu.addItem(itemMinimize, 2);
		}
		/*
		 * if (packFrame) { frame.pack(); } else { frame.validate(); }
		 * 
		 * Dimension screenSize = Toolkit.getDefaultToolkit().getScreenSize();
		 * 
		 * Dimension frameSize = frame.getSize(); if (frameSize.height >
		 * screenSize.height) { frameSize.height = screenSize.height; } if
		 * (frameSize.width > screenSize.width) { frameSize.width =
		 * screenSize.width; }
		 * 
		 * 
		 * Make the window fullscreen - On Request of users This seems not to
		 * work on sun's version 1.4.1_01 Works great with 1.4.2 !!! So update
		 * your J2RE or J2SDK.
		 */
		/* Used to maximize the screen if the JVM Version if 1.4 or higher */
		/* --------------------------------------------------------------- */
		double JVMVer =
			Double
				.valueOf(System.getProperty("java.version").substring(0, 3))
				.doubleValue();

		frame.pack();
		if (JVMVer >= 1.4) {
			frame.setExtendedState(Frame.MAXIMIZED_BOTH);
		} else {
			frame.setExtendedState(Frame.NORMAL);
		}
		/* --------------------------------------------------------------- */
		/* Added By Jeremy Whitlock (jcscoobyrs) 07-Nov-2003 at 15:54:24 */

		// Not needed ???
		frame.setVisible(true);
		frame.toFront();
		frame.requestFocus();

	}

	public static void closeWindow() {
		if (frame == null)
			return;
		frame.dispose();
		//long m1 = Runtime.getRuntime().freeMemory();
		//frame = null;
		if (!sysTrayDisabled) {
			menu.removeAll();
			menu.addItem(itemExit, 0);
			menu.addSeparator(1);
			menu.addItem(itemRestore, 2);
		} else
			System.out.println(
				"Memoranda is active in a 'daemon mode'. To restore the window, run the application again.");
		//CurrentProject.free();
		System.gc(); // ??? Are we really need it?
		//long m2 = Runtime.getRuntime().freeMemory();
		/* DEBUG */ //System.out.println("Window closed. "+(m2-m1)/1024);
	}

	/**
	 * @see snoozesoft.systray4j.SysTrayMenuListener#menuItemSelected(snoozesoft.systray4j.SysTrayMenuEvent)
	 */
	public void menuItemSelected(SysTrayMenuEvent e) {
		if (e.getActionCommand().equals("restore"))
			show();
		else if (e.getActionCommand().equals("minimize"))
			closeWindow();
		else if (e.getActionCommand().equals("exit")) {
			if (frame != null)
				frame.doExit();
			else
				System.exit(0);
		}
	}

	/**
	 * @see snoozesoft.systray4j.SysTrayMenuListener#iconLeftClicked(snoozesoft.systray4j.SysTrayMenuEvent)
	 */
	public void iconLeftClicked(SysTrayMenuEvent arg0) {
		show();
	}

	/**
	 * @see snoozesoft.systray4j.SysTrayMenuListener#iconLeftDoubleClicked(snoozesoft.systray4j.SysTrayMenuEvent)
	 */
	public void iconLeftDoubleClicked(SysTrayMenuEvent arg0) {
		show();
	}

	/**
	 * Method showSplash.
	 */
	private void showSplash() {
		splash = new JFrame();
		ImageIcon spl =
			new ImageIcon(App.class.getResource("resources/splash.png"));
		JLabel l = new JLabel();
		l.setSize(400, 300);
		l.setIcon(spl);
		splash.getContentPane().add(l);
		splash.setSize(400, 300);
		Dimension screenSize = Toolkit.getDefaultToolkit().getScreenSize();
		splash.setLocation(
			(screenSize.width - 400) / 2,
			(screenSize.height - 300) / 2);
		splash.setUndecorated(true);
		splash.setVisible(true);
	}
}